Issue an error when using `no_mangle` on language items

alexandruradovici opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This is related to https://github.com/rust-lang/rust/issues/139923
    r? @bjorn3
-->
<!-- homu-ignore:end -->

This pull requests adds the code to issue an error or a warning when using `no_mangle` on language items. This should detail why the `undefined symbol` error is issued for the code described in #139923.

The pull request adds two ui tests, one testing the error and the other one the warning.

I would love some feedback here, as I am not sure that the error and warning are issues using the right API.

alexandruradovici commented:
r? @bjorn3

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14621386262/job/41022157145) [(plain)](https://github.com/rust-lang/rust/commit/caa2b0489f2d134ac61e62d7e73b4c201cd00f53/checks/41022157145/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 19.7s done
#18 DONE 33.7s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Number of decisions:   4447
  longest path:          1159 (code:    152)
  longest backtrack:       66 (code:    428)
Shared 86733 out of 152951 states by creating 14756 new states, saving 71977
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc: In function ‘rtx_def* extract_bit_field_1(rtx, poly_uint64, poly_uint64, int, rtx, machine_mode, machine_mode, bool, bool, rtx_def**)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1864:45: warning: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ may be used uninitialized [-Wmaybe-uninitialized]
 1864 |       rtx sub = extract_bit_field_as_subreg (mode1, op0, imode,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
 1865 |                                              bitsize, bitnum);
      |                                              ~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1824:19: note: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ was declared here
 1824 |   scalar_int_mode imode;
      |                   ^~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc: In member function ‘void range_def_chain::dump(FILE*, basic_block, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc:319:19: warning: format not a string literal and no format arguments [-Wformat-security]
  319 |           fprintf (f, prefix);
      |           ~~~~~~~~^~~~~~~~~~~
---
                 from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/region-model.h:33,
                 from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/access-diagram.cc:39:
In constructor ‘ana::byte_range::byte_range(ana::byte_offset_t, ana::byte_size_t)’,
    inlined from ‘virtual text_art::table ana::string_literal_spatial_item::make_table(const ana::bit_to_table_map&, text_art::style_manager&) const’ at /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/access-diagram.cc:1812:18:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/store.h:312:5: warning: ‘size_in_bytes.generic_wide_int<fixed_wide_int_storage<128> >::fixed_wide_int_storage<128>.fixed_wide_int_storage<128>::val[1]’ may be used uninitialized [-Wmaybe-uninitialized]
  312 |     m_size_in_bytes (size_in_bytes)
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/access-diagram.cc: In member function ‘virtual text_art::table ana::string_literal_spatial_item::make_table(const ana::bit_to_table_map&, text_art::style_manager&) const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/analyzer/access-diagram.cc:1808:28: note: ‘size_in_bytes.generic_wide_int<fixed_wide_int_storage<128> >::fixed_wide_int_storage<128>.fixed_wide_int_storage<128>::val[1]’ was declared here
 1808 |                byte_size_t size_in_bytes
      |                            ^~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/diagnostic.cc: In function ‘void fancy_abort(const char*, int, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/diagnostic.cc:1677:15: warning: format not a string literal and no format arguments [-Wformat-security]
 1677 |       fnotice (stderr, diagnostic_kind_text[DK_ICE]);
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:7930:9: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 7930 |   write (fd, "\n\n", 2);
      |   ~~~~~~^~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In member function ‘void driver::final_actions() const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:9307:13: warning: ignoring return value of ‘int truncate(const char*, __off_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 9307 |     truncate(totruncate_file, 0);
      |     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-playback.h:31,
                 from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-builtins.cc:24:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h: In member function ‘virtual bool gcc::jit::recording::type::is_same_type_as(gcc::jit::recording::type*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.h:640:20: warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc: In function ‘void lto_resolution_read(splay_tree, FILE*, lto_file*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2091:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2091 |   fscanf (resolution, " ");   /* Read white space.  */
      |   ~~~~~~~^~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2093:9: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2093 |   fread (obj_name, sizeof (char), name_len, resolution);
      |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2113:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2113 |   fscanf (resolution, "%u", &num_symbols);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.cc:32:
---
Applying io_quotes_use            to linux/blkzoned.h
Applying io_quotes_use            to linux/ipmi.h
Applying io_quotes_use            to linux/psp-dbc.h
Applying io_quotes_use            to linux/bt-bmc.h
Applying io_quotes_use            to linux/tps6594_pfsm.h
Applying io_quotes_use            to linux/cxl_mem.h
Applying io_quotes_use            to linux/wmi.h
Applying io_quotes_use            to linux/auto_fs.h
Applying io_quotes_use            to linux/mmtimer.h
Applying io_quotes_use            to linux/f2fs.h
Applying io_quotes_use            to linux/vhost.h
---
Applying io_quotes_use            to sound/asound.h
Applying io_quotes_use            to sound/compress_offload.h
Applying hpux8_bogus_inlines      to math.h
Applying pthread_incomplete_struct_argument to pthread.h
Fixed:  pthread.h
Applying io_quotes_use            to misc/mrvl_cn10k_dpi.h
Applying io_quotes_use            to misc/ocxl.h
Applying io_quotes_use            to misc/cxl.h
Applying io_quotes_use            to misc/xilinx_sdfec.h
Applying io_quotes_def            to unicode/platform.h
Applying sun_malloc               to malloc.h
Applying io_quotes_use            to scsi/cxlflash_ioctl.h
---
Applying machine_name             to x86_64-linux-gnu/bits/unistd_ext.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/mtrr.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/amd_hsmp.h
Applying machine_name             to openssl/e_os2.h
Applying io_quotes_use            to drm/xe_drm.h
Applying io_quotes_use            to drm/radeon_drm.h
Applying io_quotes_use            to drm/panfrost_drm.h
Applying io_quotes_use            to drm/etnaviv_drm.h
Applying io_quotes_use            to drm/lima_drm.h
Applying io_quotes_use            to drm/qaic_accel.h
Applying io_quotes_use            to drm/vc4_drm.h
Applying io_quotes_use            to drm/i915_drm.h
Applying io_quotes_use            to drm/omap_drm.h
Applying io_quotes_use            to drm/pvr_drm.h
Applying io_quotes_use            to drm/amdgpu_drm.h
Applying io_quotes_use            to drm/vgem_drm.h
Applying io_quotes_use            to drm/msm_drm.h
Applying io_quotes_use            to drm/v3d_drm.h
Applying io_quotes_use            to drm/exynos_drm.h
Applying io_quotes_use            to drm/nouveau_drm.h
Applying io_quotes_use            to drm/drm.h
Applying io_quotes_use            to drm/habanalabs_accel.h
Applying io_quotes_use            to drm/tegra_drm.h
Applying io_quotes_use            to rdma/rdma_user_ioctl.h
cc1: note: self-tests are not enabled in this build
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc: In function ‘void server(bool, int, module_resolver*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc:620:10: warning: ignoring return value of ‘int pipe(int*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
---
+    | ^^^^^^^^^^^^
+ LL | static X: u32 = 42;
+    | ------------- is the panic_impl item
+    |
+    = note: The linker requires specific names for internal language items,
+    = note: like panic handlers, eh personality functions or eh catch typeinfo variables.
+    = note: In this case it is automatically renamed to `rust_begin_unwind`
+ 
+ error: aborting due to 3 previous errors
10 
11 For more information about this error, try `rustc --explain E0718`.
12 

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/panic-handler/panic-handler-wrong-location.rs:7:1
- LL | #[panic_handler] //~ ERROR `panic_impl` lang item must be applied to a function
+ error: `#[no_mangle]` cannot be used on weak language items
+   --> $DIR/panic-handler-wrong-location.rs:7:1
+    |
+ LL | #[panic_handler]
+    | ---------------- this defines it as the `panic_impl` item
+ LL | #[no_mangle]
+    | ^^^^^^^^^^^^
+ LL | static X: u32 = 42;
+    | ------------- is the panic_impl item
+    |
+    = note: The linker requires specific names for internal language items,
+    = note: like panic handlers, eh personality functions or eh catch typeinfo variables.
+    = note: In this case it is automatically renamed to `rust_begin_unwind`
+ 
+ error: aborting due to 3 previous errors


The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args panic-handler/panic-handler-wrong-location.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/panic-handler/panic-handler-wrong-location.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/panic-handler/panic-handler-wrong-location" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-C" "panic=abort"
stdout: none
--- stderr -------------------------------
error[E0718]: `panic_impl` lang item must be applied to a function
##[error]  --> /checkout/tests/ui/panic-handler/panic-handler-wrong-location.rs:6:1
   |
LL | #[panic_handler] //~ ERROR `panic_impl` lang item must be applied to a function
   | ^^^^^^^^^^^^^^^^ attribute should be applied to a function, not a static item

error: `#[panic_handler]` function required, but not found

error: `#[no_mangle]` cannot be used on weak language items
##[error]  --> /checkout/tests/ui/panic-handler/panic-handler-wrong-location.rs:7:1
   |
LL | #[panic_handler] //~ ERROR `panic_impl` lang item must be applied to a function
   | ---------------- this defines it as the `panic_impl` item
LL | #[no_mangle]
   | ^^^^^^^^^^^^
LL | static X: u32 = 42;
   | ------------- is the panic_impl item
   |
   = note: The linker requires specific names for internal language items,
   = note: like panic handlers, eh personality functions or eh catch typeinfo variables.
   = note: In this case it is automatically renamed to `rust_begin_unwind`

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0718`.
```

</details>

alexandruradovici commented:
This fails due to a test that defines `#[panic_handler]` in a wrong place that also defines `#[no_mangle]`.

```rust
#[panic_handler] //~ ERROR `panic_impl` lang item must be applied to a function
#[no_mangle]
static X: u32 = 42;
```

This somehow seems to be an error priority issue, the error that this pull request issues is correct.

ppannuto commented on compiler/rustc_codegen_ssa/src/codegen_attrs.rs:
```diff
@@ -608,11 +610,55 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
+                let label = format!("is the {} item", lang_item.name());
+                let mut err = tcx.dcx()
+                    .struct_span_err(no_mangle_span, "`#[no_mangle]` cannot be used on weak language items")
+                    .with_note("The linker requires specific names for internal language items,")
+                    .with_note("like panic handlers, eh personality functions or eh catch typeinfo variables.")
```
This might be too specialized, but I imagine that many folks putting `no_mangle` on internal functions are doing so for debugging purposes. It might be nice to [include the debugging advice from the original issue](https://github.com/rust-lang/rust/issues/139923#issuecomment-2815568150) in the compiler hint here, i.e.:

```suggestion
                    .with_note("The linker requires specific names for internal language items,")
                    .with_note("like panic handlers, eh personality functions or eh catch typeinfo variables.")
                    .with_note("If you are trying to prevent mangling to ease debugging, many")
                    .with_note("debuggers support a command such as `rbreak {sym}` to")
                    .with_note("match `.*{sym}.*` instead of `break {sym}` on a specific name")
```

alexandruradovici replied:
Thanks @ppannuto.

Noratrieb commented:
> This fails due to a test that defines `#[panic_handler]` in a wrong place that also defines `#[no_mangle]`.
> 
> ```rust
> #[panic_handler] //~ ERROR `panic_impl` lang item must be applied to a function
> #[no_mangle]
> static X: u32 = 42;
> ```
> 
> This somehow seems to be an error priority issue, the error that this pull request issues is correct.
> 
> I am not sure exactly how to avoid this, as I don't think checking that the target is a function is correct, as some language items are not functions.

you can probably just fix the test if you look at the history of the test and ensure that what it's trying to test is still tested

alexandruradovici commented:
Thank you @Noratrieb 

I deleted the `no_mangle` attribute from the `ui/panic-handler/panic-handler-wrong-location`. I am not sure why that was placed there, but as it now is an error, I think this should be correct. I might be wrong though, so any feedback is welcome.

bjorn3 commented on compiler/rustc_codegen_ssa/src/codegen_attrs.rs:
```diff
@@ -608,11 +610,58 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
+        let has_no_mangle = codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::NO_MANGLE);
         if WEAK_LANG_ITEMS.contains(&lang_item) {
             codegen_fn_attrs.flags |= CodegenFnAttrFlags::RUSTC_STD_INTERNAL_SYMBOL;
+            if has_no_mangle {
```
This code is only running when you are defining a lang item. I think it should prevent all combinations of `#[rustc_std_internal_symbol]` and `#[no_mangle]`. Can you move it below the `lang_items::extract` if?

alexandruradovici replied:
Just to make sure I understand this correctly, you're suggesting to add another if after the `lang_items::extract`:

```rust
if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::NO_MANGLE) {
  if is_weak_symbol {
   // issue error
  } else 
  if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::RUSTC_STD_INTERNAL_SYMBOL) {
    // issue warning
  } 
}
```

bjorn3 replied:
Looking at it a bit more careful, I think keeping the error in the same location would be fine. The warning should be moved out of the lang item if however.

Also I noticed that you didn't make the warning a future incompat warning: https://rustc-dev-guide.rust-lang.org/diagnostics.html#future-incompatible-lints

alexandruradovici replied:
> Looking at it a bit more careful, I think keeping the error in the same location would be fine. The warning should be moved out of the lang item if however.
> 
Got it.

> Also I noticed that you didn't make the warning a future incompat warning: https://rustc-dev-guide.rust-lang.org/diagnostics.html#future-incompatible-lints

Yep, I wasn't sure if this was the indented way, I'll change it. I have two questions here:
- what issue should I reference?
- what should be the reason? `FutureReleaseErrorDontReportInDeps`?

bjorn3 replied:
`FutureReleaseErrorReportInDeps` should be fine as reason. We don't need to emit a future incompat error for long given that `#[rustc_std_internal_symbol]` is unstable. As for the issue, that would be a newly created tracking issue for this future incompat warning.

bjorn3 replied:
> On the other hand, the nightly versions already issue linking errors.

Yeah. I think emitting an error immediately for this case is fine.

alexandruradovici replied:
Do you have an example of a language item that would emit a warning? I am not able to find one that would fit the criteria and I want to have a test for it.

bjorn3 replied:
I think the only lang items for which `#[no_mangle]` is actively harmful are those that are weak lang items or those that are generic. Weak lang items would be a hard error and generic lang items don't allow `#[no_mangle]` already I think.

alexandruradovici replied:
I modified the code to issue just the error.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14750542770/job/41406877502) [(plain)](https://github.com/rust-lang/rust/commit/817fe981f937f603c4863b9fc2b28d083c424d17/checks/41406877502/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.887 Building wheels for collected packages: reuse
#12 2.888   Building wheel for reuse (pyproject.toml): started
#12 3.102   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.103   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=5bb60f62728aaedff7162745ce743c7f2f55069b3e7f82e6a37d70df455797cc
#12 3.103   Stored in directory: /tmp/pip-ephem-wheel-cache-154n3nnv/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.105 Successfully built reuse
#12 3.106 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.508 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.509 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.062 Collecting virtualenv
#12 4.117   Downloading virtualenv-20.30.0-py3-none-any.whl (4.3 MB)
#12 4.310      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 22.7 MB/s eta 0:00:00
#12 4.371 Collecting filelock<4,>=3.12.2
#12 4.379   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.418 Collecting platformdirs<5,>=3.9.1
#12 4.425   Downloading platformdirs-4.3.7-py3-none-any.whl (18 kB)
#12 4.447 Collecting distlib<1,>=0.3.7
#12 4.454   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.468      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 39.6 MB/s eta 0:00:00
#12 4.549 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.734 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.7 virtualenv-20.30.0
#12 4.734 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.8s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      122816 kB
DirectMap2M:     7217152 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.216
---
fmt check
fmt: checked 6000 files
tidy check
tidy: Skipping binary file check, read-only filesystem
tidy error: /checkout/tests/ui/codegen/no-mangle-on-internal-lang-items.rs: missing trailing newline
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
Collecting pip
  Downloading pip-25.1-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-25.1-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 71.1 MB/s eta 0:00:00
Installing collected packages: pip
---
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:50
  local time: Wed Apr 30 08:58:19 UTC 2025
  network time: Wed, 30 Apr 2025 08:58:19 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bjorn3 commented on compiler/rustc_codegen_ssa/src/codegen_attrs.rs:
```diff
@@ -621,6 +623,44 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
+                no_mangle_span,
+                "`#[no_mangle]` cannot be used on internal language items",
+            )
+            .with_note("The linker requires specific names for internal language items.")
```
```suggestion
            .with_note("Rustc requires this item to have a specific mangled name.")
```

not all of them are lang items, you can also use `#[rustc_std_internal_symbol]` explicitly. And it is rustc which has the expectation on the symbol name, not the linker.

alexandruradovici commented on tests/ui/codegen/no-mangle-on-internal-lang-items.rs:
```diff
@@ -0,0 +1,14 @@
+
+fn main() {
+
+}
```
Fixed

alexandruradovici commented on compiler/rustc_codegen_ssa/src/codegen_attrs.rs:
```diff
@@ -621,6 +623,44 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
+            .with_note("match `.*{sym}.*` instead of `break {sym}` on a specific name")
+            .with_span_label(tcx.def_span(did), label);
+        if let Some(lang_item) = lang_item {
+            if let Some(span) = span {
+                let label_attr = format!("this defines it as the `{}` item", lang_item.name());
+                err = err.with_span_label(span, label_attr);
+            }
```
Is there any other way to say what language item this is? I have not seen any way of saying which language item it is.

alexandruradovici replied:
I deleted it for now.

alexandruradovici commented on compiler/rustc_codegen_ssa/src/codegen_attrs.rs:
```diff
@@ -621,6 +623,44 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
+                let label_attr = format!("this defines it as the `{}` item", lang_item.name());
+                err = err.with_span_label(span, label_attr);
+            }
+            if let Some(link_name) = lang_item.link_name() {
+                let renamed =
+                    format!("In this case it is automatically renamed to `{}`", link_name);
+                err = err.with_note(renamed);
+            }
```
Is there any way to get the mangled name? I am assuming no, as the mangling is done later on.

alexandruradovici replied:
I tried this:

```rust
let instance = Instance::mono(tcx, did.into());
let name = tcx.symbol_name(instance).name;
```

but it throws a cycle detected error.

alexandruradovici commented on compiler/rustc_codegen_ssa/src/codegen_attrs.rs:
```diff
@@ -621,6 +623,44 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
+            )
+            .with_note("The linker requires specific names for internal language items.")
+            .with_note("If you are trying to prevent mangling to ease debugging, many")
+            .with_note("debuggers support a command such as `rbreak {sym}` to")
```
I don't think @ppannuto purpose was rot replace these. In any case, we would need the mangled name. Is there any way to get it, as I am assuming that mangling is performed later on.

alexandruradovici replied:
I used the local name to replace *sym* as anyway this would be a regular expression match, but it is the name before the `#[rustc_std_internal_symbol]`

alexandruradovici commented on compiler/rustc_codegen_ssa/src/codegen_attrs.rs:
```diff
@@ -621,6 +623,44 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
+            .map_or((None, None), |(name, span)| (Some(span), LangItem::from_name(name)));
+        // This should never use the default as `no_mangle` is present in the attributes list
+        let no_mangle_span = no_mangle_span.unwrap_or_default();
+        let label = if let Some(lang_item) = lang_item {
+            format!("should be the {} item", lang_item.name())
+        } else {
+            format!("should be the internal language item")
+        };
```
I wanted a way to show the user the *problematic* item, not just the attribute.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14751104168/job/41408703714) [(plain)](https://github.com/rust-lang/rust/commit/7c977426cc14117d9612aa70b810aa6049daaaae/checks/41408703714/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 23.0s done
#18 DONE 29.3s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---

6 LL | fn internal_lang_function () {
7    | ---------------------------- should be the internal language item
8    |
-    = note: The linker requires specific names for internal language items.
+    = note: Rustc requires this item to have a specific mangled name.
10    = note: If you are trying to prevent mangling to ease debugging, many
11    = note: debuggers support a command such as `rbreak {sym}` to
12    = note: match `.*{sym}.*` instead of `break {sym}` on a specific name


The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args codegen/no-mangle-on-internal-lang-items.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/codegen/no-mangle-on-internal-lang-items.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/codegen/no-mangle-on-internal-lang-items" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2024"
stdout: none
--- stderr -------------------------------
error: `#[no_mangle]` cannot be used on internal language items
##[error]  --> /checkout/tests/ui/codegen/no-mangle-on-internal-lang-items.rs:7:1
   |
LL | #[unsafe(no_mangle)] //~ERROR `#[no_mangle]` cannot be used on internal language items
   | ^^^^^^^^^^^^^^^^^^^^
LL | fn internal_lang_function () {
   | ---------------------------- should be the internal language item
   |
   = note: Rustc requires this item to have a specific mangled name.
   = note: If you are trying to prevent mangling to ease debugging, many
   = note: debuggers support a command such as `rbreak {sym}` to
   = note: match `.*{sym}.*` instead of `break {sym}` on a specific name

error: aborting due to 1 previous error
------------------------------------------


---- [ui] tests/ui/codegen/no-mangle-on-panic-handler.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/codegen/no-mangle-on-panic-handler/no-mangle-on-panic-handler.stderr"
diff of stderr:

8 LL | pub unsafe fn panic_fmt(pi: &PanicInfo) -> ! {
9    | -------------------------------------------- should be the panic_impl item
10    |
-    = note: The linker requires specific names for internal language items.
+    = note: Rustc requires this item to have a specific mangled name.
12    = note: If you are trying to prevent mangling to ease debugging, many
13    = note: debuggers support a command such as `rbreak {sym}` to
14    = note: match `.*{sym}.*` instead of `break {sym}` on a specific name


The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args codegen/no-mangle-on-panic-handler.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/codegen/no-mangle-on-panic-handler.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/codegen/no-mangle-on-panic-handler" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2024"
stdout: none
--- stderr -------------------------------
error: `#[no_mangle]` cannot be used on internal language items
##[error]  --> /checkout/tests/ui/codegen/no-mangle-on-panic-handler.rs:10:1
   |
LL | #[unsafe(no_mangle)] //~ ERROR `#[no_mangle]` cannot be used on internal language items
   | ^^^^^^^^^^^^^^^^^^^^
LL | #[panic_handler]
   | ---------------- this defines it as the `panic_impl` item
LL | pub unsafe fn panic_fmt(pi: &PanicInfo) -> ! {
   | -------------------------------------------- should be the panic_impl item
   |
   = note: Rustc requires this item to have a specific mangled name.
   = note: If you are trying to prevent mangling to ease debugging, many
   = note: debuggers support a command such as `rbreak {sym}` to
   = note: match `.*{sym}.*` instead of `break {sym}` on a specific name
   = note: In this case it is automatically renamed to `rust_begin_unwind`

error: aborting due to 1 previous error
------------------------------------------

```

</details>

bjorn3 commented on compiler/rustc_codegen_ssa/src/codegen_attrs.rs:
```diff
@@ -621,6 +623,40 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
+        let lang_item =
+            lang_items::extract(attrs).map_or(None, |(name, _span)| LangItem::from_name(name));
+        // This should never use the default as `no_mangle` is present in the attributes list
+        let no_mangle_span = no_mangle_span.unwrap_or_default();
+        let mut err = tcx
+            .dcx()
+            .struct_span_err(
+                no_mangle_span,
```
```suggestion
        let mut err = tcx
            .dcx()
            .struct_span_err(
                no_mangle_span.unwrap_or_default(),
```

bjorn3 reviewed (approved):
r=me with the commented out code removed.

alexandruradovici committed:
error when using no_mangle on language items

add suggestion on how to add a panic breakpoint

Co-authored-by: Pat Pannuto <pat.pannuto@gmail.com>
delete no_mangle from ui/panic-handler/panic-handler-wrong-location test

issue an error for the usage of #[no_mangle] on internal language items

delete the comments

add newline

rephrase note

Co-authored-by: bjorn3 <17426603+bjorn3@users.noreply.github.com>
update error not to leak implementation details

delete no_mangle_span

Co-authored-by: bjorn3 <17426603+bjorn3@users.noreply.github.com>
delete commented code

alexandruradovici commented:
> Would you mind squashing your changes?

Done

bors commented:
:pushpin: Commit 07c7e5ffb3c096f80ffede421fb20089fbe7fe3f has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 07c7e5ffb3c096f80ffede421fb20089fbe7fe3f -->
<!-- homu: {"type":"Approved","sha":"07c7e5ffb3c096f80ffede421fb20089fbe7fe3f","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
